libs/@local/harpc/net/src/session/test.rs

```diff
@@ -0,0 +1,426 @@
+    }
+
+    let elapsed = time.elapsed();
+
+    assert_eq!(bytes, payload_len);
+
+    println!("Elapsed: {elapsed:?}");
+
+    // calculate the throughput in bytes per second
+    #[expect(
+        clippy::float_arithmetic,
+        clippy::cast_precision_loss,
+        reason = "statistics"
+    )]
+    let throughput = (payload_len as f64) / elapsed.as_secs_f64();
+    let formatter = ISizeFormatter::new(throughput, humansize::BINARY);
+
+    println!("Throughput: {formatter:.2}/s");
+}
+
+#[test_log::test(tokio::test(flavor = "multi_thread", worker_threads = 8))]
+async fn echo_memory() {
+    let address = memory_address();
+
+    echo(libp2p::core::transport::MemoryTransport::default, address).await;
+}
+
+#[test_log::test(tokio::test(flavor = "multi_thread", worker_threads = 8))]
+async fn echo_tcp() {
+    let address: Multiaddr = [
+        multiaddr::Protocol::Ip4(Ipv4Addr::LOCALHOST),
+        multiaddr::Protocol::Tcp(0),
+    ]
+    .into_iter()
+    .collect();
+
+    echo(libp2p::tcp::tokio::Transport::default, address).await;
+}
+
+struct ClientOptions {
+    length: usize,
+    index: u8,
+}
+
+async fn echo_client<const VERIFY: bool>(
+    connection: Connection,
+    ClientOptions { length, index }: ClientOptions,
+) -> Duration {
+    let descriptor = Descriptor::default();
+
+    let payload = Bytes::from(vec![index; length]);
+    let payload_len = payload.len();
+
+    let time = Instant::now();
+
+    let mut stream = connection
+        .call(
+            descriptor.service,
+            descriptor.procedure,
+            stream::iter(iter::once(payload)),
+        )
+        .await
+        .expect("connection should be open");
+
+    let mut response = stream
+        .next()
+        .await
+        .expect("should receive a response")
+        .expect("value response");
+
+    let mut bytes = 0;
+
+    while let Some(chunk) = response.next().await {
+        bytes += chunk.len();
+
+        if VERIFY {
+            assert!(chunk.iter().all(|&byte| byte == index));
+        }
+    }
+
+    let elapsed = time.elapsed();
+
+    assert_eq!(bytes, payload_len);
+
+    elapsed
+}
+
+async fn echo_concurrent<T>(
+    mut transport: impl FnMut() -> T + Send,
+    address: Multiaddr,
+    clients: u8,
+) where
+    T: Transport,
+{
+    let (server, _server_guard) = server(
+        TransportConfig::default(),
+        server::SessionConfig::default(),
+        transport(),
+    );
+    let server_ipc = server.transport().ipc().clone();
+
+    let address = {
+        let server_stream = server
+            .listen(address)
+            .await
+            .expect("should be able to listen on TCP");
+
+        // Give the swarm some time to acquire the external address
+        // This is necessary for CI, as otherwise the tests are a bit flaky.
+        // TODO: `listen_on` should wait until the transport layer has acquired said address.
+        tokio::time::sleep(tokio::time::Duration::from_millis(50)).await;
```

Comment:
```suggestion
        // Give the swarm some time to acquire the external address
        // This is necessary for CI, as otherwise the tests are a bit flaky.
        // TODO: `listen_on` should wait until the transport layer has acquired said address.
        //   see https://linear.app/hash/issue/H-2837/[bilal]-implement-waiting-for-server-to-be-ready
        tokio::time::sleep(tokio::time::Duration::from_millis(50)).await;
```